core/commands.py

```diff
@@ -0,0 +1,158 @@
+import os
+import time
+from collections import OrderedDict
+
+from django.core.cache import cache
+from django.db import connection, transaction
+from django.db.utils import ProgrammingError
+from django.utils import timezone
+from rows.utils import ProgressBar, open_compressed, pgimport
+
+from core.models import Field, Table
+
+
+class ImportDataCommand:
+    def __init__(self, table, **options):
+        self.table = table
+        self.flag_import_data = options["import_data"]
+        self.flag_vacuum = options["vacuum"]
+        self.flag_clear_view_cache = options["clear_view_cache"]
+        self.flag_create_filter_indexes = options["create_filter_indexes"]
+        self.flag_fill_choices = options["fill_choices"]
+        self.collect_date = options["collect_date"]
+
+    @classmethod
+    def execute(cls, dataset_slug, tablename, filename, **options):
+        table = Table.with_hidden.for_dataset(dataset_slug).named(tablename)
+        self = cls(table, **options)
+        db_table_suffix = "_temp"
+
+        if self.flag_import_data:
+            Model = self.table.get_model(cache=False, db_table_suffix=db_table_suffix)
+            self.import_data(filename, Model)
+            self.replace_model(TargetModel=self.table.get_model(cache=False), TempModel=Model)
+            Model = self.table.get_model(cache=False)
+        else:
+            Model = self.table.get_model(cache=False)
+        if self.flag_vacuum:
+            self.run_vacuum(Model)
+        if self.flag_create_filter_indexes:
+            self.create_filter_indexes(Model)
+        if self.flag_fill_choices:
+            self.fill_choices(Model)
+        if self.flag_clear_view_cache:
+            print("Clearing view cache...")
+            cache.clear()
+
+    def import_data(self, filename, Model):
+        # Create the table if not exists
+        with transaction.atomic():
+            try:
+                Model.delete_table()
+            except ProgrammingError:  # Does not exist
+                pass
+            finally:
+                Model.create_table(create_indexes=False)
+                Model.create_triggers()
+
+        # Get file object, header and set command to run
+        table_name = Model._meta.db_table
+        database_uri = os.environ["DATABASE_URL"]
+        encoding = "utf-8"  # TODO: receive as a parameter
+        timeout = 0.1  # TODO: receive as a parameter
+        start_time = time.time()
+        progress = ProgressBar(prefix="Importing data", unit="bytes")
+
+        # TODO: change the way we do it (CSV dialect may change, encoding
+        # etc.)
+        file_header = open_compressed(filename).readline().strip().split(",")
```

Comment:
Better to solve this TODO, so we can use any type of CSV:

```suggestion
        sample_size = 1 * 1024 * 1024  # 1 MiB
        with open_compressed(filename, mode="rb") as fobj:
            sample = fobj.read(sample_size)
            dialect = rows.plugins.csv.discover_dialect(sample, encoding)
        with open_compressed(filename) as fobj:
            reader = csv.DictReader(fobj, dialect=dialect)
            file_header = reader.fieldnames
```

Reply:
Fixed by

- 1f024247f17ffdaf69d4df0aaa05c5705e287e66
- 295c138b2a87945dbe9ae11ef01a099d0096f2a7